src/Websites/Websites/Cmdlets/WebApps/PublishAzureWebApp.cs

```diff
@@ -0,0 +1,107 @@
+using System.Net;
+using System.Net.Http;
+using System.Text;
+using System.Threading;
+
+namespace Microsoft.Azure.Commands.WebApps.Cmdlets.WebApps
+{
+    /// <summary>
+    /// Deploy a web app from a ZIP, WAR, or JAR archive.
+    /// </summary>
+    [Cmdlet("Publish", ResourceManager.Common.AzureRMConstants.AzureRMPrefix + "WebApp", SupportsShouldProcess = true, DefaultParameterSetName = ParameterSet2Name), OutputType(typeof(PSSite))]
+    public class PublishAzureWebAppCmdlet : WebAppOptionalSlotBaseCmdlet
+    {
+        // Poll status for a maximum of 20 minutes (1200 seconds / 2 seconds per status check)
+        private const int NumStatusChecks = 600;
+
+        [Parameter(Mandatory = true, HelpMessage = "The path of the archive file. ZIP, WAR, and JAR are supported.")]
+        [ValidateNotNullOrEmpty]
+        public string ArchivePath { get; set; }
+
+        [Parameter(Mandatory = false, HelpMessage = "Deploy the web app without prompting for confirmation.")]
+        public SwitchParameter Force { get; set; }
+
+        [Parameter(Mandatory = false, HelpMessage = "Run cmdlet in the background")]
+        public SwitchParameter AsJob { get; set; }
+
+        public override void ExecuteCmdlet()
+        {
+            base.ExecuteCmdlet();
+            User user = WebsitesClient.GetPublishingCredentials(ResourceGroupName, Name, Slot);
+
+            HttpResponseMessage r;
+            string deployUrl;
+            string deploymentStatusUrl = user.ScmUri + "/api/deployments/latest";
+
+            if (ArchivePath.ToLower().EndsWith("war"))
+            {
+                deployUrl = user.ScmUri + "/api/wardeploy?isAsync=true";
+            }
+            else if (ArchivePath.ToLower().EndsWith("zip") || ArchivePath.ToLower().EndsWith("jar"))
+            {
+                deployUrl = user.ScmUri + "/api/zipdeploy?isAsync=true";
+            }
+            else
+            {
+                throw new Exception("Unknown archive type.");
+            }
+
+            Action zipDeployAction = () =>
+            {
+                using (var s = File.OpenRead(ArchivePath))
+                {
+                    HttpClient client = new HttpClient();
+                    var byteArray = Encoding.ASCII.GetBytes(user.PublishingUserName + ":" + user.PublishingPassword);
+                    client.DefaultRequestHeaders.Authorization = new System.Net.Http.Headers.AuthenticationHeaderValue("Basic", Convert.ToBase64String(byteArray));
+                    HttpContent fileContent = new StreamContent(s);
+                    fileContent.Headers.ContentType = new System.Net.Http.Headers.MediaTypeHeaderValue("multipart/form-data");
+                    r = client.PostAsync(deployUrl, fileContent).Result;
+
+                    int numChecks = 0;
+                    do
+                    {
+                        Thread.Sleep(TimeSpan.FromSeconds(2));
+                        r = client.GetAsync(deploymentStatusUrl).Result;
```

Comment:
What if the result is not ready in 2 sec?

Reply:
@Nking92 Please address Vlad's comment.